shub/config.py

```diff
@@ -269,11 +271,37 @@ def get_target_conf(self, target, auth_required=True):
             apikey=apikey,
             stack=(self.stacks.get(proj['stack'], proj['stack'])
                    if 'stack' in proj else self.stacks.get('default')),
+            image=self._select_image_for_project(target, proj),
             requirements_file=self.requirements_file,
             version=self.get_version(),
             eggs=self.eggs,
         )
 
+    def _select_image_for_project(self, target, project):
+        """Helper to select image for a project (or its target).
+
+        The select logic is the following:
+        - image defined per project has highest priority,
+        - images section is marked as deprecated, but still in force:
+          if target is defined in images - use a corresponding image,
+        - if image is not defined for project or target, stacks should
+          be used instead
+
+        The function responds with a custom image name string
+        (or None meaning regular stack-based deploy).
+        """
+        if target in self.images and target != 'default':
```

Comment:
It is [always saved back as `image:` and not as `images: default:`](https://github.com/scrapinghub/shub/blob/master/shub/config.py#L169-L171) (unless there's other entries in the `images` dictionary).

But for loading, indeed, after `ShubConfig.load()` has processed the YAML, it is impossible to distinguish whether the config file contained `image:` or `images: default:`. We could move this deprecation warning to `ShubConfig.load()` though.